### PR TITLE
[TT-373] Restart Subscription

### PR DESCRIPTION
**Pull request by kalverra**
This is the first step in improving our tests' durability when faced with unreliable RPC nodes. This should cover the most basic case for now.

### RPC Issues We Gotta Handle

- [x] Handle RPC down for a short amount of time
- [x] Handle missed blocks during downtime
- [ ] Handle RPC down for a long amount of time
- [ ] Multiple RPC node failover
- [ ] Handle bad info from RPCs

### CODE REVIEW

File: blockchain/transaction_confirmers.go [hunk 1]
```diff
@@ -298,36 +299,84 @@ func (e *EthereumClient) GetHeaderSubscriptions() map[string]HeaderEventSubscrip
...
+					log.Error().Dur("Time waiting", time.Since(rpcDegradedTime)).Msg("RPC connection still down, timed out waiting for it to come back up")
+					e.Client.Close()
+					return
+				case <-ticker.C: // Poll the RPC connection to see if it's back up
+					if time.Since(rpcDegradedNotifyTime) >= time.Second*15 { // Periodically inform that we're still waiting for RPC to come back up
+						log.Warn().Dur("Time waiting", time.Since(rpcDegradedTime)).Msg("RPC connection still down, waiting for it to come back up")
+						rpcDegradedNotifyTime = time.Now()
+					}
+					subscription, err = e.SubscribeNewHeaders(context.Background(), headerChannel)
+					if err == nil { // No error on resubscription, RPC connection restored, back to regularly scheduled programming
```
Comment (AnieeG) [inline blockchain/transaction_confirmers.go:339]: do we need to handle the else part?

Comment (kalverra) [reply blockchain/transaction_confirmers.go:339]: We do, we just log a trace that we failed that attempt and go back to the beginning of the loop and wait for the next tick.

File: blockchain/transaction_confirmers.go [hunk 2]
```diff
@@ -298,36 +299,84 @@ func (e *EthereumClient) GetHeaderSubscriptions() map[string]HeaderEventSubscrip
 	return newMap
 }
 
-// subscribeToNewHeaders
+// subscribeToNewHeaders kicks of the primary header subscription for the test loop
```
Comment (tateexon) [inline blockchain/transaction_confirmers.go:302]:
```suggestion
// subscribeToNewHeaders kicks off the primary header subscription for the test loop
```

**Review by tateexon — APPROVED** [review]

File: blockchain/transaction_confirmers.go [hunk 3]
```diff
@@ -298,36 +299,84 @@ func (e *EthereumClient) GetHeaderSubscriptions() map[string]HeaderEventSubscrip
...
 			subscription.Unsubscribe()
 
-			subscription, err = e.SubscribeNewHeaders(context.Background(), headerChannel)
-			if err != nil {
-				log.Error().Err(err).Msg("Failed to resubscribe to new headers")
-				return err
+			rpcDegradedTime, rpcDegradedNotifyTime := time.Now(), time.Now()
+			timeout, ticker := time.NewTimer(e.NetworkConfig.Timeout.Duration), time.NewTicker(time.Second)
+		reSubLoop:
+			for { // Loop to resubscribe to new headers,
```
Comment (AnieeG) [inline blockchain/transaction_confirmers.go:327]: nit: for better readability can you please create another function for the resub

File: blockchain/transaction_confirmers.go [hunk 4]
```diff
@@ -298,36 +299,84 @@ func (e *EthereumClient) GetHeaderSubscriptions() map[string]HeaderEventSubscrip
...
+func (e *EthereumClient) backfillMissedBlocks(lastBlockSeen uint64, headerChannel chan *SafeEVMHeader) error {
+	latestBlockNumber, err := e.LatestBlockNumber(context.Background())
+	if err != nil {
+		return err
+	}
+	if latestBlockNumber <= lastBlockSeen {
+		return nil
+	}
+	for i := lastBlockSeen + 1; i <= latestBlockNumber; i++ {
+		header, err := e.HeaderByNumber(context.Background(), big.NewInt(int64(i)))
```
Comment (AnieeG) [inline blockchain/transaction_confirmers.go:373]: I am a bit worried on repetitive rpc calls. Can we use [SubscribeNewHead](https://github.com/ethereum/go-ethereum/blob/master/ethclient/ethclient.go#L329) ?

**Comment by kalverra** [comment]
After doing some soak testing over the weekend, this mostly works well, but it seems there are some serious issues (maybe performance?) with the backfill. Investigating.

**Comment by kalverra** [comment]
Will close this for now and re-open when settled.

**Pull request closed without merging**